migrate client sdk including sis into ddc-sdk

gharibyan opened the pull request:
1. migrated client sdk with sis initial implementation from activity sdk
2. fixed examples test run.

gharibyan committed:
migrate client sdk including sis into ddc-sdk

gharibyan committed:
fix type check and tests.

gharibyan committed:
update docker composer images.

gharibyan committed:
implement quic support publishing in sis

gharibyan committed:
fix tests added sis browser playground.

gharibyan committed:
fix node version.

gharibyan committed:
Update package.json

gharibyan committed:
fix build configs.

gharibyan committed:
fix client sdk packaging.

spoluyan commented on packages/client/src/types.ts:
There are duplicate type definitions between `packages/client/src/types.ts` and `packages/client/src/sis/types.ts`. 
Both files define `ServiceUnavailableError`, `DataStream`, `Packet`, `HandshakeRequest`, etc. with different shapes:

- `types.ts`: `HandshakeRequest.type is 'publish' | 'subscribe' (string)`
- `sis/types.ts`: `HandshakeRequest.type is number (0 or 1)`

Consolidate all types into a single source of truth.

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,116 @@
+import MPC from './mcp';
+import Wallet from './wallet';
+
+const DEFAULT_WALLET_CONFIG = 'hybrid label reunion only dawn maze asset draft cousin height flock nation';
```
Remove the default wallet entirely and require explicit wallet configuration, or throw an error if no wallet is provided.

gharibyan replied:
Done.

spoluyan commented on packages/client/src/context.ts:
```diff
@@ -0,0 +1,15 @@
+export class ClientContext implements ContextInterface {
+  public agent_service: string;
+  public workspace: string;
+  public domain?: string;
```
Must be `stream`, not `domain`. Also, this field is required.

gharibyan replied:
Done!

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,116 @@
+    publisher: async (streamId: string) => {
+      return this.sis.newPublisher(streamId);
+    },
+    subscribe: (streamId: string, callback: (data: { headers: Packet['headers']; data: any }) => void) => {
```
Potential memory leak.

Issues:

- The async IIFE is "fire-and-forget" with no way to unsubscribe
- No error handling if the subscription fails
- No return value to allow cleanup

spoluyan commented on packages/client/src/sis/http.ts:
```diff
@@ -0,0 +1,179 @@
+        }
+        if (hasJson) {
+          try {
+            const j = await (response as any).json();
```
The code calls `response.text()` first, then later tries to call `response.json()`. Once the body is consumed by `text()`, `json()` will fail. This logic is convoluted and error-prone.

spoluyan commented on playground/src/BrowserStreaming.tsx:
```diff
@@ -0,0 +1,561 @@
+
+  // Subscriber state
+  const [subscribed, setSubscribed] = useState<boolean>(false);
+  const [offsetInput, setOffsetInput] = useState<string>('');
```
The offset input field is rendered but never used when subscribing. Maybe file for Playground.

spoluyan commented on packages/client/src/sis/types.ts:
```diff
@@ -0,0 +1,279 @@
+export class SISError extends Error {
+  constructor(
+    message: string,
+    public code?: string,
```
Minor - many places throw SISError without a code.

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,116 @@
+      const errMessage = (json && json.error && json.error.message) || `Request failed with status ${response.status}`;
+      const error = new Error(errMessage);
+      // @ts-expect-error augment error with extra context
+      error.status = response.status;
```
Maybe it makes sense Instead of using @ts-expect-error, create a proper typed error class.

spoluyan commented on packages/client/src/sis/transport.ts:
```diff
@@ -0,0 +1,452 @@
+      await quicheLoaded;
+      return WebTransport;
+    } catch (e) {
+      console.log(e);
```
`console.log(e);` leaks into production. Use proper logging infrastructure.

spoluyan commented on package.json:
Not for this PR, but we need to fix Polkadot dependencies. The package declares dependencies on Polkadot utilities but doesn't specify them as peer dependencies, which could lead to duplicate versions. Maybe just add TODO for now.

gharibyan replied:
added todo in readme

spoluyan commented on packages/client/package.json:
```diff
@@ -0,0 +1,49 @@
+    "@cere-activity-sdk/ciphers": "0.1.7",
+    "@cere-activity-sdk/signers": "0.1.7",
+    "@cere/embed-wallet": "^0.23.0",
+    "@fails-components/webtransport": "^1.4.8",
```
Listed twice (once in root `package.json`, once here). May cause version conflict in future.

spoluyan commented on packages/client/src/sis/client.ts:
```diff
@@ -0,0 +1,474 @@
+   * @throws SISError if configuration is invalid
+   */
+  constructor(config: ClientConfig) {
+    this.config = {
```
Maybe move default values to constants?

spoluyan commented on packages/client/src/sis/client.ts:
```diff
@@ -0,0 +1,474 @@
+  /**
+   * Closes all connections and cleans up resources.
+   */
+  async close(): Promise<void> {
```
There's no way to gracefully shut down `ClientSdk`. The `SisClient` has a `close()` method but `ClientSdk` doesn't expose it.

gharibyan committed:
fix pr comments.

1. fixed types now sis/types.ts and client/types.ts are explicit belongs to each.
2. fixed memory leak on subscription
3. added unsubscribe method.
4. fixed package duplication comments.

gharibyan commented:
@spoluyan tests are failing because of linter. if i run it at the moment there will be lot's of file changes. i'll fix that and testing flow with different pr.

gharibyan committed:
add packetParsing support to process json, binary and text.

gharibyan committed:
fix linting, fix failed test

gharibyan committed:
update version

gharibyan committed:
fix package lock.

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,129 @@
+    publisher: async (streamId: string) => {
+      return this.sis.newPublisher(streamId);
+    },
+    subscribe: (
```
Problems:

- The async IIFE runs independently with no cancellation mechanism
- Calling unsubscribe() doesn't stop the loop — it only closes the SIS client
- The loop continues even after unsubscribe() is called
- Multiple subscriptions create multiple uncancellable loops

Impact: memory leaks, resource waste, inability to stop subscriptions
Recommendation: return an AbortController or a cancellation function.

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,129 @@
+        }
+      })();
+    },
+    unsubscribe: () => {
```
Problems:

- this.sis.close() closes all transports and HTTP clients
- If a user has both a publisher and subscriber, calling unsubscribe() will close both
- No way to unsubscribe from a single subscription without affecting other operations

Impact: breaks concurrent usage (publish + subscribe), unexpected connection drops
Recommendation: track individual subscriptions and close only the specific one

spoluyan commented on packages/client/src/sis/http.ts:
```diff
@@ -0,0 +1,134 @@
+          throw new ServiceUnavailableError();
+        }
+        if (response.status === 404) {
+          throw new SISError('Not found', 'NOT_FOUND', 404);
```
The HTTP client defines `StreamNotFoundError` but throws a generic SISError for 404

spoluyan commented on packages/client/src/utils.ts:
```diff
@@ -0,0 +1,24 @@
+
+  if (contentType.includes('application/json')) {
+    const text = new TextDecoder().decode(packet.payload);
+    data = JSON.parse(text);
```
The first `JSON.parse` (line 10) can throw if the payload is invalid JSON. The second one (line 16) is wrapped in try/catch. Make both consistent.

spoluyan commented on packages/client/src/sis/transport.ts:
```diff
@@ -0,0 +1,385 @@
+    this.stream = await this.transport.createStream();
+    const writer = this.stream.writable.getWriter();
+    this.reader = this.stream.readable.getReader();
+    this.abortController = new AbortController();
```
Created but never used to abort operations. Either use it or remove it.

gharibyan committed:
update changes based on pr review.

gharibyan committed:
Update RecordPage.tsx

gharibyan committed:
update pr review changes.

gharibyan committed:
update eventRuntimePath

gharibyan committed:
fix playground build

gharibyan committed:
fix nodejs support packaging issue.

gharibyan committed:
fix @fails-components/webtransport bundling dependency.

gharibyan committed:
update version

gharibyan committed:
add cubby query in ddc sdk.

gharibyan committed:
fix blakejs import on es build

gharibyan committed:
add logging in certeficate hash.

spoluyan commented on packages/client/src/sis/certificate.ts:
SDK should not own logging; callers should plug in a logger.

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,143 @@
+      const url = this.buildURL(this.agentRuntimeUrl, path);
+      const requestBody = {
+        params: payload,
+      } as CubbyQueryRequestBody;
```
The type says `payload` but the code sends `params`. The cast hides the mismatch.

spoluyan commented on packages/client/src/types.ts:
```diff
@@ -0,0 +1,43 @@
+    public code?: string | number,
+  ) {
+    super(message);
+    this.name = 'CubbyRequestError';
```
`name` is set to `'CubbyRequestError'` while the class is `CubbyError`. That breaks the usual expectation that `err.name === 'CubbyError'` and can confuse logging/monitoring.

spoluyan commented on packages/client/src/client.ts:
```diff
@@ -0,0 +1,143 @@
+      if (timeoutMs) {
+        requestBody.timeoutMs = timeoutMs;
+      }
+      const response = await fetch(url, {
```
No `signal` or timeout. Long-running or stuck requests cannot be cancelled and can hang indefinitely.

gharibyan committed:
remove certificate logs, fix cubbies queries timeout and error handling.
